src/main/java/com/cloudbees/dockerpublish/DockerBuilder.java

```diff
@@ -49,24 +51,25 @@
 
     private static final Pattern IMAGE_BUILT_PATTERN = Pattern.compile("Successfully built ([0-9a-f]+)");
 
-    private final String repoName;
-    private final boolean noCache;
-    private final boolean forcePull;
-    private final String dockerfilePath;
-    private final boolean skipBuild;
-    private final boolean skipDecorate;
+    @CheckForNull
+    private final DockerServerEndpoint server;
+    @CheckForNull
+    private DockerRegistryEndpoint registry;
```

Comment:
Alternately, _do_ allow both to be null, do not give them initializer values, do not try to set them in `readResolve` except to the extent that `registry.url` can be inferred from an old `repoName`, have their `@DataBoundSetter`s replace blank values (`server.url == null && server.credentialsId == null`) with nulls, and have the getters replace null with a blank value. That keeps the XStream form cleaner.